protoc_docs/bin/py_desc_converter.py

```diff
@@ -0,0 +1,199 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+import sys
+
+import pypandoc
+
+from google.protobuf import descriptor_pb2 as desc
+
+
+class CommentsBuilder(object):
+    """Comments builder which converts comments in a batch by calling
+    convert_text only once for the whole batch. The ``pypandoc.convert_text``
+    call spawns a child ``pandoc`` subprocess, so it is an expensive operation
+    and we want to minimize the number of such calls.
+
+    Batching is performed by concatenating comments with special characters in
+    them (and skipping the others) using the unique batch token. After that
+    ``pandoc`` is called for the batch only once and then the result is split by
+    same batch token back into individual comments.
+    """
+
+    _PROTO_LINK_RE = re.compile(
+        r'(\[(?P<text>[^\]]+)\]\[(?P<uri>[A-Za-z_][A-Za-z_.0-9]*)?\])')
+    _RELATIVE_LINK_RE = re.compile(
+        r'(\[(?P<text>[^\]]+)\]\((?P<uri>/[^\)]+)\))')
+    _NEW_LINES = re.compile(r'(?P<newlines>(\r?\n)+)(?P<followup>[^\r\n])')
+
+    _BATCH_TOKEN = "$#!"
+
+    def __init__(self):
+        self.raw_comments = {}
+        self.converted_comments = {}
+        self._index = 0
+
+    def put_comment(self, comment):
+        """Put a comment in a batch for future processing by ``pypandoc``.
+
+        Args:
+            comment (str): The comment to convert.
+        """
+
+        comment = self._replace_proto_link(comment)
+        comment = self._replace_relative_link(comment)
+
+        token = comment
+
+        # Try to avoid conversion for comments without special characters in the
+        # markdown
+        if any([i in token for i in '`[]*_']):
+            if len(self.converted_comments) > 0:
+                token = "\n%s%s" % (self._BATCH_TOKEN, token)
+            self.converted_comments[self._index] = token
+        else:
+            self.raw_comments[self._index] = token
+
+        self._index += 1
+
+    def build(self):
+        """Builds the comments by calling `pypandoc` for a batch of comments and
+        then splitting them back into individual comments.
+        """
+
+        converted_doc = pypandoc.convert_text(
+            "".join(self.converted_comments.values()),
+            'rst',
+            format='commonmark'
+        )
+        converted = converted_doc.split(self._BATCH_TOKEN)
+
+        c_index = 0
+        index = 0
+        while c_index < len(converted):
+            while index in self.raw_comments:
+                index += 1
+            self.converted_comments[index] = self._insert_spaces(
+                converted[c_index])
+            index += 1
+            c_index += 1
+
+        self._index = 0
+
+    def get_next_comment(self):
```

Comment:
This method has the implicit precondition that it's only called the correct number of times. It feels a little unsafe, even though it's clearly correct in the below usage.

Reply:
You are absolutely right about the precondition and that it is unsafe... If it was a larger scale system I would probably avoid doing it, but for a smaller script seems like an ok thing to do. Otherwise we will have to introduce more complexity  here than it needs.

I'll add more comments describing this aspect. Should it be enough, WDYT?

Reply:
I think just adding comments are fine for this particular case.